CIMPL6 Import

cmoesel opened the pull request:
Changes needed to support CIMPL6.  The two most significant changes to the CIMPL6 model are:

* Replaced `code`, `Coding`, and `CodeableConcept` with new `concept` primitive
* Removed `RefValue` in favor of using a field's "entry-ness" to determine if it should be a reference

To support the new `concept` primitive (and removal of the other code-ish classes), I created a `Concept` class that is output with (and used by) the generated code.

All unit tests should pass.  For information on running w/ other required modules and CIMPL6 source see:
https://github.com/standardhealth/shr-cli/pull/207

Marking as a DRAFT PR because all versions will need to be bumped to non-beta versions upon approval.

dehall commented on lib/includes/json-helper.js:
```diff
@@ -217,8 +218,12 @@ export const FHIRHelper = {
+   *
+   * @param {Object} shrObject - the SHR object to reference
+   * @param {List<Object>} referenceList - the list of objects that are referenced
+   * @returns {string} the reference
```
Should be `{Reference}` not string. That's really my bad for not commenting the function in the first place. Same thing in `createReferenceWithoutObject` below

cmoesel replied:
Well, now that I look at it, it's pretty obvious from the code -- so really, it's my bad.  Either copy/paste error or I just made a bad assumption without verifying it.  I'll fix it.

dehall commented on lib/includes/Coding.js:
```diff
@@ -0,0 +1,93 @@
+ * A coding representation.  This should only be used in the context of a Concept.
+ */
+export default class Coding {
+  constructor(code, system, display) {
```
I was surprised when I saw these using the constructor since none of the generated classes do and instead use `fromJSON`/field setters, but then I saw the Reference class does the same thing. I don't think it's worth changing now (unless you think it can be done really quickly & easily), but at some point I think we may want to try to make the Concept/Coding/Reference classes work the same way as the generated classes.

cmoesel replied:
We can look into it.  It's a little tricky for `concept` since the JSON passed in might be a `code`, a `Coding`, or a `Concept` -- but I think I already have to figure that out elsewhere anyway, so I guess it would be just moving that logic into here.

I think I was just thinking that from the point of view of a user, creating instances in code (not from JSON), having a simple constructor is easier than having to use a default constructor and then set those things individually.

cmoesel commented on lib/includes/Concept.js:
```diff
@@ -0,0 +1,44 @@
+   * @param {Array<Coding>} coding - The coding array
+   * @returns {Concept} this.
+   */
+  withCodings(coding) {
```
Oh shoot.  Yes.  I originally had `Codings` (with an `s`) all around.  I then changed it to be `Coding` so it's consistent w/ everything else, but must have missed this one.  Will fix it.

cmoesel commented on lib/generateClass.js:
```diff
@@ -540,6 +545,8 @@ function writeFromFhirProfile(def, specs, fhir, fhirProfile, cw) {
   const allFhirElements = preProcessChoiceFields(fhirProfile.snapshot.element, specs, fhir);
   preProcessFieldMappings(allFhirElements, specs, def);
 
+  // Determine isQuantity outside of the iteration loop so it is more efficient.  See note below for reason we need isQuantity.
```
I'm not sure exactly what you mean by "recognize this kind of situation more generically", but... consider this an interim approach for now.  I think where we want to go is to allow the components of `concept` (even though it is technically a primitive) to be mapped just like other things.  So instead of:
```
map Units to code
```
(which is what we have now and is problematic), we can have something more like this:
```
map Units[concept].code to code
map Units[concept].system to system
map Units[concept].display to unit
```

This will be more like everything else and should hopefully require very little special case code (just to deal, generally, with being able to map components of `concept`).

cmoesel commented:
@dehall -- I've addressed your comments (well, the short-term ones anyway).  We tried to shield most of the changes from the rest of the framework, but the replacement of `CodeableConcept`, `Coding`, and `code`, as well as the removal of `RefValue`, was pretty much impossible to happen transparently.  So... here we are.

It's worth noting that Mark has changed a lot of the base classes and such as well, so I don't think the grammar change is going to be the only challenge...
